src/lib/support/ThreadOperationalDataset.cpp

```diff
@@ -0,0 +1,629 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <assert.h>
+#include <string.h>
+
+#include "ThreadOperationalDataset.h"
+
+namespace chip {
+namespace Thread {
+
+/**
+ * This class implements Thread Tlv functionality.
+ *
+ */
+class Tlv
+{
+    static constexpr uint8_t kLengthEscape = 0xff; ///< This length value indicates the actual length is of two-bytes length.
+
+public:
+    enum : uint8_t
+    {
+        kChannel         = 0,
+        kPanId           = 1,
+        kExtendedPanId   = 2,
+        kNetworkName     = 3,
+        kPSKc            = 4,
+        kMasterKey       = 5,
+        kMeshLocalPrefix = 7,
+        kActiveTimestamp = 14,
+    };
+
+    uint8_t GetSize(void) const { return static_cast<uint8_t>(sizeof(mType) + sizeof(mLength) + GetLength()); }
+
+    /**
+     * This method returns the Tlv type.
+     *
+     * @returns The Tlv type.
+     *
+     */
+    uint8_t GetType(void) const { return mType; }
+
+    /**
+     * This method sets the Tlv type.
+     *
+     */
+    void SetType(uint8_t aType) { mType = aType; }
+
+    /**
+     * This method returns the Tlv length.
+     *
+     * @returns The Tlv length.
+     *
+     */
+    uint8_t GetLength(void) const
+    {
+        assert(mLength != kLengthEscape);
+        return mLength;
+    }
+
+    /**
+     * This method sets the length.
+     */
+    void SetLength(uint8_t aLength)
+    {
+        assert(mLength != kLengthEscape);
+        mLength = aLength;
+    }
+
+    /**
+     * This method returns a pointer to the value.
+     *
+     * @returns The Tlv value.
+     *
+     */
+    const void * GetValue(void) const
+    {
+        assert(mLength != kLengthEscape);
+
+        return reinterpret_cast<const uint8_t *>(this) + sizeof(mType) + sizeof(mLength);
```

Comment:
This memory manipulation does not seem safe - assumes that the class is derived and is followed by a suitably sized byte buffer.

sizeof(this) would be probably more accurate than sizeof(mType) + sizeof(mLength). However still, there seems no way for this class to validate available buffer lengths. It seems more approriate to use existing TLV reader/writer classes in CHIP.